### DISCUSSION TITLE
Auto-merging of backporting PRs

### DISCUSSION THREAD

**Discussion by FlorianHockmann** [JanusGraph/janusgraph#3460] [General]
#3297 enabled auto-merging of backporting PRs. My assumption was that this would only let the tool merge the PR after all actions were executed and all checks turned green. However, it looks like it only waits for _required_ checks, but we currently don't require any checks for a PR to be merged. The reason why no checks are formally required is probably that we have a lot of flaky tests and some checks are prone for FPs like the test coverage or Codacy.

I would like to discuss here how we want to proceed with auto-merging of backporting PRs. I see three options:
1. Disable auto-merge completely.
2. Keep auto-merge as is without any required checks.
3. Make some/most of our checks _required_ so the auto-merge will wait for them.

Option 1 (disable auto-merge) means of course more manual work for us as we need to check the backporting PRs and merge them after all checks have passed.
Option 2 (auto-merge without checks) requires basically no effort from us. It could maybe also be argued that running all checks again on `v0.6` isn't necessary if they have already passed on `master`. This is probably also similar to our previous merging strategy where we have merged `v0.6` into `master` locally after merging a PR into `v0.6`. Most of us probably also haven't executed all tests locally on `master` before pushing the branch.
Option 3 (add required checks & use them for auto-merge) seems like the safest option to me, but it changes also how we merge other PRs as they also need to pass all required checks. So, this requires more work from us.

After writing this, I personally tend towards option 3. I think it would help us in general to decide on which checks need to pass (like DCO, EasyCLA & all builds) and then motivate us to improve the situation with flaky tests. Flaky tests already slow us down a lot in my opinion (e.g. for Dependabot PR) so improving this helps us in general.
Auto-merging also makes sense in general from my side for backporting as we have already reviewed those changes for `master` so I don't see a point in reviewing them again for `v0.6`.

I would however definitely also be fine with option 1 or 2.

Any thoughts on this?

**Comment by farodin91** [comment]
I thought it using the https://github.blog/changelog/2021-02-04-pull-request-auto-merge-is-now-generally-available/ this feature which should for the merge requirements. Or could we merge, if requirements are not matched? I will later check our merge conditions.

**Reply by farodin91** [reply]
I added some basic requirements. https://github.com/JanusGraph/janusgraph/settings/branches

**Reply by FlorianHockmann** [reply]
Looks like adding those checks as requirements leads to problems for PRs that only change the docs as those checks are skipped then. This is blocking the merge for such PRs (example: #3409):

![grafik](https://user-images.githubusercontent.com/21224336/213128007-527a1483-362c-40e0-bb3a-ff0d34294ba5.png)

This however seems to be a common problem: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/troubleshooting-required-status-checks#handling-skipped-but-required-checks

The suggestion there is to create a copy of the required workflows that is executed for PRs where the normal workflow will be skipped and that always turns green. While this solution doesn't sound really nice to maintain, I think it's still the best solution for us right now.
I'll create a PR that creates such dummy workflow doubles so we can still use required checks.

**Comment by porunov** [comment]
I would vote for solution 3 if we can fix doc-only PRs as you suggested here (https://github.com/JanusGraph/janusgraph/discussions/3460#discussioncomment-4715470).
Solution 1 is OK to me. It does have requirement for manual work, but that work is straightforward and quite simple.
Solution 2 could be dangerous in cases when some dependency updated in `master` but the same update leads to conflicts in `0.6` or other branches.
Solution 3 looks like the best balance between Solution 1 and Solution 2. It may stimulate us to re-write some flaky tests better.
Basically, I like Solution 3 the most, I'm OK with Solution 1, and I'm not sure about Solution 2 but don't have a strong opinion on that.

**Reply by FlorianHockmann** [reply]
Completely forgot to link the PR here, but #3468 should fix the problem with doc-only PRs.

**Comment by porunov** [comment]
@FlorianHockmann , @farodin91 I'm disabling this requirement:
> Require branches to be up to date before merging
> This ensures pull requests targeting a matching branch have been tested with the latest code. This setting will not take effect unless at least one status check is enabled (see below).

I think it's unrealistic requirement for active branches. I.e. whenever someone merges anything we will need to trigger all branches to rebase and re-test. Tests may take quite some time and then the branch after being finished with testing may become outdated again and it will require re-testing again (i.e. endless process). For example, due to too many busy workers, the next PR #3268 was running tests for about 18 hours. After finishing with tests it again requires rebasing and testing against newest `master` commit (like doc changes or so). If I trigger tests again they will take another 18 hours or so (because we don't have enough available workers right now). After that I'm sure the branch will become outdated again.

**Reply by FlorianHockmann** [reply]
Agreed. I think we had that requirement always enabled but it simply didn't do anything as we didn't have any required checks. After @farodin91 added some required checks, the requirement suddenly kicked in which resulted in this problem.
Ultimately, it would be great of course if every PR would be tested with the current state on the target branch, but I agree that it's not practical for now to enforce that.
